keps/sig-api-machinery/2309-idl/README.md

```diff
@@ -0,0 +1,636 @@
+
+### Syntax
+
+See [the attached proposal][syntax] for grammar, syntax design, and
+in-depth examples.
+
+[syntax]: ./syntax.md
+
+### Output Scope
+
+The intial scope of this proposal is to generate two concrete artifacts:
+**CRDs** (i.e. what we have in controller-tools today) and **Go types**
+(i.e. generate types.go).
+
+The format is intended to be flexible enough that we can eventually port
+all of the generators in code-generator to make use of the IDL.  These are
+considered follow-up work after the MVP.
+
+### Concrete KubeBuilder & k/k Changes
+
+* **All types in k/k are rewritten to use KDL**.  All Go types in k/k are
+  generated.
+   
+  This guides core k/k towards KDL usage, and enables third-parties to
+  interact with k/k types well from KDL.
+
+* **The KubeBuilder project is updated to support KDL, and the pseudo-IDL
+  frontend for controller-tools is deprecated.**  This guides community
+  members towards using KDL and writing well-structured APIs.
+
+  The internal representation for controller-tools will be updated to be
+  CDKL.
+
+  To avoid breaking existing workflows, the pseudo-IDL frontend will
+  continue to be available-but-deprecated in controller-tools for an
+  extended period of time (at minimum the equivalent period of time to
+  deprecating and removing a stable feature from k/k).
+
+* **Go-to-KDL migration tooling is made available** to assist both k/k and
+  the community in migrating from pseudo-IDL to KDL.
+
+  For k/k, it's expected that this will take place in 2 phases -- an
+  automated migration that produces technically-correct-but-not-detailed
+  KDL, followed small-scale updates to bring in the richer declarative
+  features from KDL that currently exist only in imperative code (e.g.
+  converting unions to be identified as such, bringing in declarative
+  defaulting & validation).
+
+### Architecture
+
+* Humans write KDL files (see [the syntax overview][syntax]), which
+  compile to CKDL files (which are machine-readable, serialized as proto
+  blobs)
+
+* CKDL files are consumed by generators to produce source code, CRD YAML,
+  etc
+
+* Tooling for CKDL files may define custom extensions (markers) in special
+  KDL files, which then get serialized as proto Any messages to be picked
+  up by downstream tooling (this gives each marker a concrete type and
+  ensures that markers are explicitly imported so that all markers in
+  a file are known to the parser).
+
+![Architecture Diagram](./architecture.svg)
+
+### Risks and Mitigations
+
+* **Learning curve**: any new setup has a potential learning curve.  That
+  being said, the syntax should look at least familiar to folks who have
+  worked with languages that use  `item: type` syntax (e.g. Rust,
+  TypeScript, Python type hints).  Furthermore, tooling to convert between
+  Go pseudo-IDL and KDL should allow easy comparison for existing API
+  authors.
+
+* **Tooling maintenance**: this introduces several new tools (KDL
```

Comment:
Can we get a section that lists all of the tools that will be supporting the IDL?

Reply:
sure, good idea :+1: